### ISSUE TITLE
incoming buffer underrun

### ISSUE THREAD

**Issue by caedmon** [pilwon/node-ib#8]
`Incoming.process` can dispatch a message before all the data for the message has been received.  The message in this case handler ends up reading past the end of this._dataQueue.  The current message is corrupted and what's worse, when the rest of the message is received Incoming tries to start reading a new message from the middle of the last message.

I saw this frequently when using reqHistoricalData.

**Comment by heberallred** [comment]
@hklsk Do you have a specific code snippet of what you did to fix the issue?

**Comment by ajpierce** [comment]
I also have the same problem when using reqHistoricalData -- I can get it to happen reliably within an hour if i'm requesting 7 instruments at the same time (along with account summaries and positions).

Thank you both @hklsk @heberallred for letting me know I'm not crazy!  I will see if I can find a fix and publish a pull request

**Comment by ajpierce** [comment]
All right team, I'm going to need some help with this one.

# Replicating
I was able to replicate the issue reliably with the following request parameters:

```
barSize: '1 min',
durationString: '2 D',
pairs: [
        'AUDUSD',
        'EURUSD',
        'GBPUSD',
        'USDCAD',
        'USDCHF',
        'USDJPY'
    ],
```

The `endDateTime` is always set to whenever the request is made, and we request 2 days backwards.  Every minute on the minute, we make 6 of these requests simultaneously with the forex contracts specified above.

# Failed Solutions
I tried three different things based on what I read above, and none of them seemed to fix the problem.

## Attempt 1: Don't push to the buffer unless it's empty.
In `lib/socket.js`, I made the following change to the [`_onData` function on line 39](https://github.com/pilwon/node-ib/blob/master/lib/socket.js#L39):

```javascript
  // Enqueue tokens only if queue is empty; prevent buffer underruns
  var self = this;
  function enqueueData() {
    if(self._controller._incoming.isQueueEmpty()) {
        return self._controller._incoming.enqueue(tokens);
    } else {
        console.log('Queue not empty, not enqueuing data');
        self._controller._incoming.process();
        return setTimeout(enqueueData, 1000);
    }
  }
  enqueueData();
```

I thought that calling `process()` instead of `enqueue()` if the buffer wasn't empty might solve the problem, but it did not. In fact, the [`process` function](https://github.com/pilwon/node-ib/blob/master/lib/incoming.js#L1205) has logic that attempts to empty out the buffer (line 1209).

## Attempt 2: Change how enqueue works
If there is some kind of problem with data getting enqueued, I figured changing how we enqueue data might solve the problem.  It didn't, because what I attempted was functionally equivalent to what the code already does.  I was grasping at straws.

In `lib/incoming.js', I changed the [`enqueue` function](https://github.com/pilwon/node-ib/blob/master/lib/incoming.js#L1201) to:

```javascript
  if(this._dataQueue.length > 0) {
    console.log('Appending to dataQueue');
    this._dataQueue = this._dataQueue.concat(tokens);
  } else {
    console.log('dataQueue is empty, setting dataQueue = tokens');
    this._dataQueue = tokens;
  }
```

## Attempt 3: Only process data if queue is empty
This was my attempt to really understand what @hklsk was saying:
> If the buffer is 0, then pass to the incoming process. It can resolve the problem

With that in mind, I tried to change `_onData_` in `lib/socket.js` again:

```javascript
Socket.prototype._onData = function (data) {
  var tokens = data.toString().split(EOL).slice(0, -1);

  this._controller.emit('received', tokens.slice(0), data);

  // Process data queue
  var shouldProcessData = this._controller._incoming.isQueueEmpty();   // NEW LINE
  this._controller._incoming.enqueue(tokens);

  if (this._neverReceived) {
    this._controller._serverVersion = parseInt(this._controller._incoming.dequeue(), 10);
    this._controller._serverConnectionTime = this._controller._incoming.dequeue();

    this._controller.emit('server', this._controller._serverVersion, this._controller._serverConnectionTime);
  }

  // NEW BLOCK  -- Only process the data if the queue is already empty -- NEW BLOCK //
  if(shouldProcessData) {
    this._controller._incoming.process();
  } else {
      console.log('Queue was not empty, NOT firing another process command!!');
  }
  // -- END NEW BLOCK -- //

  // Async
  if (this._waitingAsync) {
    this._waitingAsync = false;
    this._controller.resume();
  }

  this._neverReceived = false;
};
```

Again, this fails to resolve the issue.

# Ideas?
So that's what I've tried, and it hasn't worked yet.  Anybody have any ideas for fixing this buffer underrun error?  This seems like a show-stopper, because it's a race condition that *could* happen at any buffer size.

Let's keep the dialog open on this one!

**Comment by hklsk** [comment]
My attempt also failed and have bugs.
My attempt is determin the last packet byte of the packet is 0x00, if 0x00 dequeue and process, else enqueue, but there sometimes the last packet byte is 0x00 and the packet is not yet totally received. The IB will separate the one return packet in 2 packets (sometimes) because the packets is too long.
I think the bug is caused by the chunked packet.
For example, when we request historical data, there should be one pakcet return, but sometimes IB may divide the packet in 2 parts [Packet 1a + Packet 1b] and there is not way to determine IB separate the packet or not, there aren't any ways to determine when the packet ended.

**Comment by lewisdawson** [comment]
@ajpierce @pilwon @hklsk In the `Socket.prototype._onData` function, try moving the received event emission to after the enqueue has completed as follows:

```
Socket.prototype._onData = function (data) {
  var tokens = data.toString().split(EOL).slice(0, -1);

  // Process data queue
  this._controller._incoming.enqueue(tokens);

  this._controller.emit('received', tokens.slice(0), data);
```

**Comment by ajpierce** [comment]
@lewisdawson Thanks for the suggestion!  Looks like we still see the same behavior with that change, unfortunately :(  Output is below.

In the process of debugging, I'm outputting the data streams to a log.  Here's what the gap looks like between messages (assume a lot of text before and after these blocks, and ignore the `info:`, because that's logging output. Newlines added by hand because they appear on one line):

```
-120151207  01:32:000.733170.733210.733170.73319-1-1.00000false
-120151207  01:33:000.733190.733220.733130.73313-1-1.00000false
-120151207  01:34:000.733130.733130.733090.73312-1-1.00000false
-120151207  01:35:000.7

// Application thinks all data is received, processes data ...

info: 33120.733120.732890.73289-1-1.00000false
-120151207  01:36:000.732890.732930.732850.73289-1-1.00000false
-120151207  01:37:000.732890.732960.732780.73296-1-1.00000false
-120151207  01:38:000.732960.733040.732960.73302-1-1.00000false
-120151207  01:39:000.733020.733320.733000.73332-1-1.00000false
...
```

You can see the format of the historical data messages is supposed to be [reqId, time, open, high, low, close, volume, count, WAP, hasGaps](https://www.interactivebrokers.com/en/software/api/apiguide/java/historicaldata.htm). Volume and count are -1 because these are Forex historical prices, WAP is 1.000000, and hasGaps is false.  My guess is that the reqId is -1 if they are bars following the previous request.  

Between these two log messages, the application thinks we've received data (it's not waiting for the rest of the data to come in), and we run into our race condition.

To print this output, I added a `data.toString()` log statement to the top of the `_onData` function.

**Comment by lewisdawson** [comment]
@ajpierce Do you have a small test program you can post or are you testing all this with your main application?

**Comment by ajpierce** [comment]
@lewisdawson I've gone ahead and created [a gist that replicates the issue](https://gist.github.com/ajpierce/60f216142143b34764cf). It assumes you're using a cloned version of `node-ib` rather than the one from npm, and that you're running this script from the parent directory.

I should have done this earlier, thank you for asking!  Hopefully we can get more people checking out this bug :+1:

**Comment by lewisdawson** [comment]
@ajpierce Thanks for posting that! I'll have a look at it over the next week when I have some free time and see if I can track down a viable solution.

**Comment by quantverse** [comment]
Hello,
is this issue relevant only to reqHistoricalData? Or it can cause troubles also in other situations?
I am about to start building an app which requires to retrieve real-time FX quotes, post FX orders, process execution events,  get account stats and positions. No need for historical data whatsoever.

If it can cause troubles in situations mentioned above too, I should probably not use this library for production code trading live account. 

Tell me your opinion please, thanks.

**Comment by owap** [comment]
@quantverse The issue is with how the application receives data. IB decides how to split up large payloads, and I suppose it's not impossible that they would split up a small payload as well.  It's not just with reqHistoricalData, it's with any request over a certain size. reqHistoricalData just happens to be the most reliable way to get a massive amount of data to replicate the issue.

I've used a number of IB libraries before, and this one is the easiest to use by far. Javascript's asynchronous nature is a natural fit for IB's API, and it will save you time compared to implementing solutions in other languages.  It also works great with RxJS and the [Reactive Programming](https://gist.github.com/staltz/868e7e9bc2a7b8c1f754) paradigm.

My vote would be: rather than to implement a solution for another library, help us fix this one issue with this one :)

**Comment by quantverse** [comment]
@owap OK :)

**Comment by ajpierce** [comment]
@lewisdawson Not sure if you've had any time to poke at this one or not, but I'm leaning towards just working around the issue by requesting multiple smaller chunks of data with the hope that IB doesn't split them up (and therefore avoid the buffer underrun error).

We should be able to buffer all the bars we get from multiple requests on a single [Observable](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/gettingstarted/exploring.md#observable--observer) sequence to end up with the large chunk of data we were looking to request initially.

The downside, of course, is that you have to use more API calls to get the same amount of data (works against your "max requests per 10 minutes" historical data limitation), but the upside is that you don't have to track down tricky bugs in open source projects :)

Let me know if you're close to a breakthrough on this!

**Comment by heberallred** [comment]
I have not followed this discussion closely (been very busy with other stuff), but I thought my latest code had fixed the buffer underrun error.  Have you tried it?

 

-Heber

**Comment by ajpierce** [comment]
@heberallred you are the MAN!  Tested your changes locally and it appears to work like a charm! I've submitted a pull request with your changes here: https://github.com/pilwon/node-ib/pull/20

Thanks a lot for sharing your fix with the community! :+1:

**Comment by heberallred** [comment]
Thanks :)  I'm glad it worked.  It took a few days for me to dissect the code and find out what was going on, but was very happy when I finally found what was causing the issue.

**Comment by pilwon** [comment]
@ajpierce's PR containing @heberallred's fix along with some cleanups was published as v0.1.7 :tada:
